imlib/draw: Add anti aliasing in draw_line.

xue-singtown opened the pull request:
Hi, I add anti aliasing feature in draw_line method. I tested Binary, Grayscale and RGB565 format.

Reference code: https://gist.github.com/randvoorhies/807ce6e20840ab5314eb7c547899de68

test code in OpenMV Cam:
```
import sensor
import time

sensor.reset()  # Reset and initialize the sensor.
sensor.set_pixformat(sensor.RGB565)# Set pixel format to RGB565 (or GRAYSCALE)
#sensor.set_pixformat(sensor.GRAYSCALE)
sensor.set_framesize(sensor.QVGA)  # Set frame size to QVGA (320x240)
sensor.skip_frames(time=2000)  # Wait for settings take effect.
while True:
    clock.tick()  # Update the FPS clock.
    img = sensor.snapshot()  # Take a picture and return the image.
    img.draw_line(10,10,100,200, (255,0,0),5)
```
Results:

after_anti_aliasing:
![after_anti_aliasing](https://github.com/openmv/openmv/assets/156400698/41c9352d-fcc8-4d92-814a-e01b3d25eec5)

no_anti_aliasing:
![no_anti_aliasing](https://github.com/openmv/openmv/assets/156400698/efe608c2-e183-4bf9-8a9a-6687df2e31e1)

iabdalkader commented:
Hi! Thank you for this contribution.

Please try to match the coding style used for `if`, `switch`, functions etc.. it might help to run `../tools/codeformat.sh` first on the file you changed. Also note we don't use camel/upper case functions, so `setPixelAA` should be:

```C
static void imlib_set_pixel_antialiased(image_t *img, int x, int y, float err, int c) {
```

It would be nice if this feature can be made configurable somehow via an option in `imlib_config.h`, for smaller/slower microcontrollers.

xue-singtown commented:
Simple performance test in OpenMV H7 Plus:

|  draw_line | thickness 1  | thickness 5 |
|---|---|---|
|  antialiased draw  | 250us  |  250us |
| point_fill draw  |  42us  |  250us |

kwagyeman commented on src/omv/boards/ARDUINO_GIGA/imlib_config.h:
```diff
@@ -88,6 +88,9 @@
 #define IMLIB_ENABLE_GET_SIMILARITY
 
+// Enable draw antialiased
+#define IMLIB_ENABLE_DRAW_ANTIALIASED
```
Change to IMLIB_ENABLE_DRAW_LINE_ANTIALIASED for all of these in the library.

kwagyeman reviewed (commented):
@xue-singtown Thanks for this work!

I left a whole bunch of comments. Note, I will turn our code formater check on. You need to make sure your code passes this.

Also, while I see that you coded this using floats. I think these algorithms actually do everything more or less with integer math. You should be able to switch to ints for all the variables (use abs() versus fabsf() in this case). This should speed up the code significantly.

Finally, we have a line clipping function. Can you add that before the draw line call to clip the line segment in length to prevent it from taking infinite time if an infinitely long line was passed?

Finally, the old draw_line function when used for thick lines is actually very bad. It wastes a whole lot of time re-drawing the same pixels when a thickness is used. This new method should completely replace it. The IMLIB_ENABLE_DRAW_ANTIALIASED should just speed up the code by skipping drawing antialised stuff. However, if you switch things over to ints the performance difference may be near zero which would remove the need to even add this ifdef. Please examine that.

kwagyeman commented:
Commit messages must be in the form:

imlib/draw: Add anti aliasing in draw_line.

Note the module/file with a colon and then period at the end.

xue-singtown commented:
I changed the float op to int. It speed up.

|  draw_line | thickness 1  | thickness 5 |
|---|---|---|
|  antialiased draw float  | 250us  |  250us |
| antialiased draw int  |  140us  |  170us |
| point_fill draw  |  42us  |  250us |

I added IMLIB_ENABLE_DRAW_THIN_LINE_ANTIALIASED in imlib_config.h.
Slower microcontrollers will call origin draw thin line.

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+    int new_c;
+
+    switch (img->pixfmt) {
+        case PIXFORMAT_BINARY:
+        {
+            new_c = c;
+            break;
+        }
```
```suggestion
        case PIXFORMAT_BINARY: {
            new_c = c;
            break;
        }
```

kwagyeman replied:
```suggestion
        case PIXFORMAT_BINARY: {
            new_c = c;
            break;
        }
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+            new_c = c;
+            break;
+        }
+        case PIXFORMAT_GRAYSCALE:
+        {
+            new_c = (((old_c - c) * err) >> 8) + c;
+            break;
+        }
```
```suggestion
        case PIXFORMAT_GRAYSCALE: {
            new_c = (((old_c - c) * err) >> 8) + c;
            break;
        }
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+            new_c = (((old_c - c) * err) >> 8) + c;
+            break;
+        }
+        case PIXFORMAT_RGB565:
+        {
+            int old_r8 = COLOR_RGB565_TO_R8(old_c);
+            int old_g8 = COLOR_RGB565_TO_G8(old_c);
+            int old_b8 = COLOR_RGB565_TO_B8(old_c);
+
+            int r8 = COLOR_RGB565_TO_R8(c);
+            int g8 = COLOR_RGB565_TO_G8(c);
+            int b8 = COLOR_RGB565_TO_B8(c);
+
+            int new_r8 = (((old_r8 - r8) * err) >> 8) + r8;
+            int new_g8 = (((old_g8 - g8) * err) >> 8) + g8;
+            int new_b8 = (((old_b8 - b8) * err) >> 8) + b8;
+
+            new_c = COLOR_R8_G8_B8_TO_RGB565(new_r8, new_g8, new_b8);
+            break;
+        }
```
```suggestion
        case PIXFORMAT_RGB565: {
            int old_r8 = COLOR_RGB565_TO_R8(old_c);
            int old_g8 = COLOR_RGB565_TO_G8(old_c);
            int old_b8 = COLOR_RGB565_TO_B8(old_c);

            int r8 = COLOR_RGB565_TO_R8(c);
            int g8 = COLOR_RGB565_TO_G8(c);
            int b8 = COLOR_RGB565_TO_B8(c);

            int new_r8 = (((old_r8 - r8) * err) >> 8) + r8;
            int new_g8 = (((old_g8 - g8) * err) >> 8) + g8;
            int new_b8 = (((old_b8 - b8) * err) >> 8) + b8;

            new_c = COLOR_R8_G8_B8_TO_RGB565(new_r8, new_g8, new_b8);
            break;
        }
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+            new_c = COLOR_R8_G8_B8_TO_RGB565(new_r8, new_g8, new_b8);
+            break;
+        }
+        default:
+        {
+            // This shouldn't happen, at least we return a valid memory block
+            return;
+        }
```
```suggestion
        default: {
            // This shouldn't happen, at least we return a valid memory block
            return;
        }
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+// https://gist.github.com/randvoorhies/807ce6e20840ab5314eb7c547899de68#file-bresenham-js-L381
+static void imlib_draw_thin_line(image_t *img, int x0, int y0, int x1, int y1, int c) {
+    /* draw a black (0) anti-aliased line on white (255) background */
+    const int dx = fabs(x1 - x0);
```
```suggestion
    const int dx = abs(x1 - x0);
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+    /* draw a black (0) anti-aliased line on white (255) background */
+    const int dx = fabs(x1 - x0);
+    const int sx = x0 < x1 ? 1 : -1;
+    const int dy = fabs(y1 - y0);
```
```suggestion
    const int dy = abs(y1 - y0);
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+
+    for (;;) {
+        /* pixel loop */
+        imlib_set_pixel_aa(img, x0, y0, 256 * fabsf(err - dx + dy) / ed, c);
```
```suggestion
        imlib_set_pixel_aa(img, x0, y0, 256 * abs(err - dx + dy) / ed, c);
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+void imlib_draw_line(image_t *img, int x0, int y0, int x1, int y1, int c, int th) {
+
+    line_t line = {x0, y0, x1, y1};
+    if (lb_clip_line(&line, 0, 0, img->w, img->h)) {
+        x0 = line.x1;
+        y0 = line.y1;
+        x1 = line.x2;
+        y1 = line.y2;
+    }
```
```suggestion
    if (!lb_clip_line(&line, 0, 0, img->w, img->h)) {
        return;
    }

    x0 = line.x1;
    y0 = line.y1;
    x1 = line.x2;
    y1 = line.y2;
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+    point_fill(img, x1, y1, -r, r, c); /* add round at end*/
+
+    /* plot an anti-aliased line of width th pixel */
+    const int ex = fabs(x1 - x0);
```
```suggestion
    const int ex = abs(x1 - x0);
```

kwagyeman commented on src/omv/imlib/draw.c:
```diff
@@ -88,27 +88,183 @@ static void point_fill(image_t *img, int cx, int cy, int r0, int r1, int c) {
+    /* plot an anti-aliased line of width th pixel */
+    const int ex = fabs(x1 - x0);
+    const int sx = x0 < x1 ? 1 : -1;
+    const int ey = fabs(y1 - y0);
```
```suggestion
    const int ey = abs(y1 - y0);
```

kwagyeman commented:
@xue-singtown - Switching to using 256 versus 255 to make the alpha blending code easier for the compiler was very clever.

I added some changes to the PR you should make. Please let me know what the speed difference is between having

IMLIB_ENABLE_DRAW_THIN_LINE_ANTIALIASED 

Enabled and disabled between 1 and 5 pixel lines. If it's not large we can then remove the ifdef and always use the new code.

xue-singtown commented:
|  IMLIB_ENABLE_DRAW_THIN_LINE_ANTIALIASED | thickness 1  | thickness 5 |
|---|---|---|
| define |  130us  |  170us |
| undefine |  42us  |  170us(antialiased) |

Do I need to use ifdef to preserve 42us thin line aliased?

kwagyeman commented:
Did you add my suggestion? That should have slightly improved performance. Can't tell from the GitHub app @xue-singtown

kwagyeman commented:
@xue-singtown - Yeah, remvoe the ifdef. We don't need to preserve the old one. The speed difference won't really matter in the long run. Please post the final numbers before I approve the PR.

xue-singtown commented:
I tested in OpenMV H7 Plus:

draw line 1 thickness: 130us
draw line 5 thickness: 170us

code:

```
import sensor, image
import pyb

sensor.reset()
sensor.set_pixformat(sensor.RGB565)
sensor.set_framesize(sensor.QVGA)
sensor.skip_frames(time = 2000)

while(True):
    img = sensor.snapshot()
    start = pyb.micros()
    img.draw_line(10,10,100,200, (255,0,0),1)
    end = pyb.elapsed_micros(start)
    print('draw line 1', end)
    
    start = pyb.micros()
    img.draw_line(10,10,100,200, (255,0,0),5)
    end = pyb.elapsed_micros(start)
    print('draw line 5', end)

```

kwagyeman commented:
@iabdalkader Approved

iabdalkader commented:
@kwagyeman @xue-singtown Looking good, thanks for the review and fixes! 

I just have one last comment, can you please replace the `/* */` style of comments with `//` ? We don't use `/* */` especially for single line comments. Also note, comments start with a capital letter. One example:

```diff
-     /* draw a black (0) anti-aliased line on white (255) background */
+    // Draw a black (0) anti-aliased line on white (255) background
```

After that it's good to merge.
